File: tools/src/DynastyTools/Models/Person.cs
namespace DynastyTools.Models;

/// <summary>
/// GM-only character prep. Mirrors the <c>GmNotes</c> interface in the Angular app.
/// The tools read and write this faithfully; only the app's player projection strips it.
/// </summary>
public sealed class GmNotes
{
    public List<string>? PersonalityTraits { get; set; }
    public string? PhysicalDescription { get; set; }
    public List<string>? Allies { get; set; }
    public List<string>? Enemies { get; set; }
    public string? Motivations { get; set; }
    public string? Goals { get; set; }
    public string? Ambitions { get; set; }
    public string? Secrets { get; set; }
}

/// <summary>
/// One person in a world file.
/// </summary>
/// <remarks>
/// <see cref="Visibility"/> is a plain string rather than an enum on purpose: a
/// validator has to be able to <em>report</em> an unrecognised value, which is
/// impossible if deserialization throws on it first.
/// </remarks>
public sealed class Person
{
    public string Id { get; set; } = string.Empty;
    public string Name { get; set; } = string.Empty;
    public string House { get; set; } = string.Empty;
    public string Nation { get; set; } = string.Empty;
    public string? Title { get; set; }
    public int? BirthYear { get; set; }
    public int? DeathYear { get; set; }
    public bool IsAlive { get; set; }
    public List<string> ParentIds { get; set; } = [];
    public List<string> SpouseIds { get; set; } = [];
    public int? SuccessionOrder { get; set; }
    public string? Notes { get; set; }
    public List<string>? Tags { get; set; }
    public int Generation { get; set; }
    public string Visibility { get; set; } = "known";
    public bool HideParentage { get; set; }
    public GmNotes? GmNotes { get; set; }
}

/// <summary>Shape of <c>data/{world}.json</c>.</summary>
public sealed class WorldPeopleFile
{
    public List<Person> People { get; set; } = [];
}

File: frontend/src/app/models/person.ts
/**
 * Whether an entity exists at all in the player-facing view.
 * `hidden` entities are absent from the player projection entirely.
 */
export type Visibility = 'known' | 'hidden';

/**
 * GM-only character prep.
 *
 * Never included in any player-facing projection under any circumstances.
 * This is independent of `visibility` / `hideParentage`: even a fully `known`
 * person's `gmNotes` stay GM-only.
 */
export interface GmNotes {
  personalityTraits?: string[];
  physicalDescription?: string;
  /** Free text or other person ids. */
  allies?: string[];
  /** Free text or other person ids. */
  enemies?: string[];
  motivations?: string;
  goals?: string;
  ambitions?: string;
  /** The GM's own prep notes on plot secrets, distinct from `visibility`. */
  secrets?: string;
}

export interface Person {
  id: string;
  name: string;
  house: string;
  nation: string;
  title?: string;
  birthYear?: number;
  deathYear?: number;
  isAlive: boolean;
  /** 0–2 entries. */
  parentIds: string[];
  /** Supports multiple marriages. */
  spouseIds: string[];
  successionOrder?: number;
  notes?: string;
  tags?: string[];
  generation: number;
  visibility: Visibility;
  /**
   * If true, this person is visible to players but their `parentIds` are not
   * (secret/rumoured parentage plots).
   */
  hideParentage: boolean;
  gmNotes?: GmNotes;
}

/**
 * A person as seen by players.
 *
 * `gmNotes` is structurally absent rather than merely blanked, so handing a
 * `PlayerPerson` to a player-facing surface cannot leak GM prep even by
 * accident — the property does not exist on the type.
 */
export type PlayerPerson = Omit<Person, 'gmNotes'>;

/**
 * The minimum shape the tree layout, overlap filter, and person card need.
 *
 * Both `Person` and `PlayerPerson` satisfy this, so every rendering path can be
 * generic over it and stays correct in either view mode.
 */
export interface PersonLike {
  id: string;
  name: string;
  house: string;
  nation: string;
  title?: string;
  birthYear?: number;
  deathYear?: number;
  isAlive: boolean;
  parentIds: string[];
  spouseIds: string[];
  successionOrder?: number;
  generation: number;
  visibility: Visibility;
  hideParentage: boolean;
}
